synapse/http/matrixfederationclient.py

```diff
@@ -714,6 +725,26 @@ async def _send_request(
                             response.code,
                             response_phrase,
                         )
+                    elif (
+                        response.code in (307, 308)
+                        and follow_redirects
+                        and response.headers.hasHeader("Location")
+                    ):
+                        # The Location header *might* be relative so resolve it.
+                        location = response.headers.getRawHeaders(b"Location")[0]
+                        new_uri = urllib.parse.urljoin(request.uri, location)
+
+                        return await self._send_request(
+                            attr.evolve(request, uri=new_uri, generate_uri=False),
```

Comment:
it's spelled `_generate_uri` in the struct definition. Is this some attrs way of declaring a private attribute that you can initialise via the constructor?

Reply:
> Is this some attrs way of declaring a private attribute that you can initialise via the constructor?

Yes, pretty much. attrs strips the preceding underscores to let you have 'private' attributes that can be initialized.

Reply:
https://www.attrs.org/en/stable/init.html#private-attributes FTR